[SC] Added a description for value transfer recovery (#421)
klaytn/klaytn-docs: docs/node/service-chain/references/value-transfer.md

> For Service Chain, there are two operator accounts: parent chain bridge operator account, service chain bridge operator account. Each operator account is used to sign transactions.

Reviewer: ```suggestion
For ServiceChain, there are two operator accounts: parent chain bridge operator account, service chain bridge operator account. Each operator account is used to sign transactions.
```

This section explains how to check the bridge contract status.

In a bridge contact, there are two nonces, `requestNonce` and `handleNonce`.
Unlike in-chain transactions, the sub-bridge can handle a higher nonce request before the lower ones.
- requestNonce : the number of user's "cross-chain value transfer" requests made to this bridge contract.
- handleNonce : the highest nonce that the sub-bridge handled.
- lowerHandleNonce : the lowest nonce that the sub-bridge should handle.

Therefore, if nonces are updated as follows, we can say the cross-chain value-transfers are processed correctly.
- "handleNonce" and "lowerHandleNonce" of the parent chain bridge contract keep approaching to the "requestNonce" of the child chain bridge contract.
- "handleNonce" and "lowerHandleNonce" keep approaching to the "requestNonce" of the parent chain bridge contract.

If "handleNonce" equals to the "requestNonce" of the counterpart bridge contract, and the "lowerHandleNonce" is greater than "handleNonce" by 1, then users' requests were all processed.

### Log <a id="log"></a>

> Below is a typical log output from SCN during normal operation.

Reviewer: ```suggestion
Below is a typical log output from a SCN during normal operation.
```

function requestValueTransfer(uint256 _uid, address _to) external
```

> # Value Transfer Recovery

Reviewer 2: @hyunsooda Out of curiosity,
 - Can bridge operators get pending transactions or removed pending transactions?
 - Do pending transactions can be recovered even though the bridge node is restarted?

Author: 1. First of all, there's no way in contract call. With console, it's possible with `subbridge.txPending` command. There's no queryable command when a transaction is removed from bridge txpool, even no push to logs. There's an add in my [preceding PR](https://github.com/klaytn/klaytn/pull/1295/files#diff-1623298e4e46f32d0a4b35ea8ba2fd0a9de08103caf3820b7ce69293b6e765e5R337)  

2. By default, all the pending transactions are automatically resent again until their receipt arrives from the counterpart chain. Thus, value transfer recovery is not necessary in this case. Say, however, a node was down somehow for mins. During that downtime, all the requests are not processed. Suppose the node is loaded again with the value transfer recovery option. In that case, all the failed transactions are imported by linear search using the different information (`requestNonce`) between my chain and the counterpart chain.

Reviewer 2: Thank you for the kind explanation!

> Value transfer request may be failed with seom reasons. Say you requested klay transfer from subbridge to mainbridge or mainbridge to subbridge.

Reviewer 3: ```suggestion
Value transfer request may be failed with some reasons. Assume that you requested KLAY transfer from subbridge to mainbridge or mainbridge to subbridge.
```

> Value transfer request may be failed with seom reasons. Say you requested klay transfer from subbridge to mainbridge or mainbridge to subbridge.

Reviewer: ```suggestion
Value transfer request may be fail for a number of reasons. Say you requested KLAY transfer from subbridge to mainbridge or mainbridge to subbridge.
```

> Value transfer request may be fail for a number of reasons. Say you requested KLAY transfer from subbridge to mainbridge or from mainbridge to subbridge.

Reviewer 3: ```suggestion
Value transfer request may be fail for a number of reasons. For example, you requested a KLAY transfer from the subbridge to the mainbridge or from the mainbridge to the subbridge.
```

Author: `Say` or `Suppose` seems nice to give an example circumstance.

Reviewer 3: `Say` looks too colloquial to me. I changed it to 'suppose'

```suggestion
Value transfer request may be fail for a number of reasons. Suppose that you requested a KLAY transfer from the subbridge to the mainbridge or from the mainbridge to the subbridge.
```

> In that case the birdge contract in the receiver side must have enough klay than requested amount of klay. If not, the transfer would be slitenly failed without error notification by return value.

Reviewer 3: ```suggestion
In that case the bridge contract in the receiver side must have enough KLAY than requested amount of KLAY. If not, the transfer would be silently failed without error notification by return value.
```

You may need to have a spell checker ;)

I'm not sure we're saying that "silently failed", because actually at this time, we don't know whether the account has insufficient KLAY or not.

Author: The pre-condition is located at the first sentence for the failure reason. Is `we don't know whether the account has insufficient KLAY or not.` not the same with the first statement?

`sligntely failed` means that the request is actually failed without obvious error to developers.

Reviewer 3: @hyunsooda 
Is there any situation that even if we fail the transaction, it is treated as it is succeeded? For not enough KLAY situation, I thought that it will retry the value transfer with the recovery logic. So, actually it is not "failed", but it is not completed.

Author: `Not completed` and `failed` have similar semantics in this context, but a careful reader may have confusion. The fact is that the initial request is failed. The recovery just resends the failed requests again and again. Strictly speaking, I'm not sure about using the keyword `not completed` seems better or not rather than `failed`.

Reviewer 3: ```suggestion
In that case the bridge contract in the receiver side must have enough KLAY than requested amount of KLAY. If not, the transfer would be silently failed without error notification by return value. You can find this failure in the log: 
```

Okay, then what about this? I'm not sure we currently have a log when this happens. Please check this once and put the error log pattern here so that anybody can easily find the error in the log.

Author: Currently, there's no way to track the log where the transfer fails. I added logs for an [observable value transfer flow](https://github.com/klaytn/klaytn/pull/1326). Please check it out.

Reviewer 3: @hyunsooda Okay. I see. Then just fix the typos.

Author: Applied already.

> In that case the birdge contract in the receiver side must have enough klay than requested amount of klay. If not, the transfer would be slitenly failed without error notification by return value.

Reviewer: ```suggestion
In that case, the bridge contract on the receiver side must have enough KLAY than the requested amount of KLAY. If not, the transfer would fail without error notification in the return value.
```

> A feature of value transfer recovery finds unhandled events and insert them into event pool again by every given interval, which means the failed transaction can be succeed again when the counterpart bridge can successfully handle that event.

Reviewer 3: ```suggestion
A feature of value transfer recovery finds unhandled events and inserts them into the event pool again by every given interval, which means the failed transaction can be succeed again when the counterpart bridge can successfully handle that event.
```

> A feature of value transfer recovery finds unhandled events and insert them into event pool again by every given interval, which means the failed transaction can be succeed again when the counterpart bridge can successfully handle that event.

Reviewer: ```suggestion
A feature of value transfer recovery allows you to find unhandled events and insert them into event pool again in a given interval, which means failed transaction can succeed again when the counterpart bridge can successfully handle that event.
```

Author: It's good too. There's no actual user interaction. This change seems to make developers feel they should do something with this. Please advise me if this change does not cause confusion as semantically.

> A feature of value transfer recovery finds unhandled events and insert them into event pool again in a given interval, which means the failed transaction can be succeed again when the counterpart bridge can successfully handle that event.

Reviewer 3: ```suggestion
The value transfer recovery logic finds unhandled events and inserts them into the event pool again in a given interval, which means the failed transaction can be succeeded again when the counterpart bridge can successfully handle that event.
```

Author: `logic` seems ambiguous since its wording makes sense anywhere. I think to keep it seems nice as it is for the abstract description level.

Reviewer 3: ```suggestion
The value transfer recovery finds unhandled events and inserts them into the event pool again in a given interval, which means the failed transaction can be succeeded again when the counterpart bridge can successfully handle that event.
```

I just removed the 'logic'. Take a look again.

> In the case of above example, the failed transaction would be eventually handled by value transfer recovery when the counterpart bridge has enough klay.

Reviewer 3: ```suggestion
In the case of above example, the failed transaction would be eventually handled by value transfer recovery when the counterpart bridge has enough KLAY.
```

> In the case of above example, the failed transaction would be eventually handled by value transfer recovery when the counterpart bridge has enough klay.

Reviewer: ```suggestion
In case of the above example, the failed transaction would be eventually handled by value transfer recovery when the counterpart bridge has enough KLAY.
```

Author: Updated.

> In case of the above example, the failed transaction would be eventually handled by value transfer recovery when the counterpart bridge has enough KLAY.

Reviewer 3: ```suggestion
In case of the above example, the failed transaction would be eventually handled by the value transfer recovery when the counterpart bridge has enough KLAY later.
```

> In order to set the value transfer recovery as default, you need to set two properties:

Reviewer 3: ```suggestion
In order to set the value transfer recovery by default, you need to set two properties below:
```

```
SC_VTRECOVERY=1
SC_VTRECOVERY_INTERVAL=5
```

> The value transfer recovery runs automatically by set `SC_VTRECOVERY=1`. `SC_VTRECOVERY_INTERVAL` means an interval how often the value transfer recovery is executed.

Reviewer 3: ```suggestion
The value transfer recovery runs automatically by setting `SC_VTRECOVERY` to `1`. `SC_VTRECOVERY_INTERVAL` indicates an interval how often the value transfer recovery is executed.
```

Author: `SC_VTRECOVERY=1` seems more intuitive rather than an oral description. The others are okay to me.

Reviewer 3: ```suggestion
The value transfer recovery runs automatically by the statement `SC_VTRECOVERY=1`. `SC_VTRECOVERY_INTERVAL` indicates an interval how often the value transfer recovery is executed.
```

Then, what about this?

set `SC_VTRECOVERY=1` seems weird to me.

Author: How about `by setting the SC_VTRECOVERY=1`?. `statement` seems awkward wording.

Reviewer 3: ```suggestion
The value transfer recovery runs automatically by `SC_VTRECOVERY=1`. `SC_VTRECOVERY_INTERVAL` indicates an interval how often the value transfer recovery is executed.
```

Then, what about just using the assignment itself?
@terrikwak it would be also good to help from TW.

Author: This change seems sounds good to me.

> The value transfer recovery runs automatically by set `SC_VTRECOVERY=1`. `SC_VTRECOVERY_INTERVAL` means an interval how many often to be executed the value transfer recovery.

Reviewer 3: ```suggestion
The value transfer recovery runs automatically by set `SC_VTRECOVERY=1`. `SC_VTRECOVERY_INTERVAL` means an interval how often the value transfer recovery is executed.
```

> The value transfer recovery runs automatically by set `SC_VTRECOVERY=1`. `SC_VTRECOVERY_INTERVAL` means an interval how many often to be executed the value transfer recovery.

Reviewer: ```suggestion
The value transfer recovery runs automatically by setting `SC_VTRECOVERY=1`. `SC_VTRECOVERY_INTERVAL` refers to the interval for how often the value transfer recovery should be executed.
```

Author: Updated.

# Collecting Fee for KLAY/ERC-20 transfer <a id="collecting-fee-for-klay-erc-20-transfer"></a>

> In Service Chain, there is a fee collecting feature for the KLAY/ERC-20 transfer.

Reviewer: ```suggestion
In ServiceChain, there is a fee collecting feature for KLAY/ERC-20 transfers.
```

> **It will be updated soon.**

Reviewer: ```suggestion
**To be updated soon.**
```

# Customizing your Bridge Contract  <a id="customizing-your-bridge-contract"></a>

> In Service Chain, you can use your own customized Bridge contract that inherits from the original Bridge contract for your own unique service.

Reviewer: ```suggestion
In ServiceChain, you can use your own customized Bridge contract that inherits from the original Bridge contract for your own unique service.
```
